relays/bin-substrate/src/cli/chain_schema.rs

```diff
@@ -0,0 +1,494 @@
+#[derive(StructOpt, Debug, PartialEq, Eq, Clone, Copy, EnumString, EnumVariantNames)]
+pub enum RuntimeVersionType {
+	/// Auto query version from chain
+	Auto,
+	/// Custom `spec_version` and `transaction_version`
+	Custom,
+	/// Read version from bundle dependencies directly.
+	Bundle,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+pub struct RuntimeVersionParams {
+	pub prefix: &'static str,
+
+	pub version_mode: RuntimeVersionType,
+	pub spec_version: Option<u32>,
+	pub transaction_version: Option<u32>,
+}
+
+impl RuntimeVersionParams {
+	/// Converts self into `ChainRuntimeVersion`.
+	pub fn into_runtime_version(
+		self,
+		bundle_runtime_version: Option<sp_version::RuntimeVersion>,
+	) -> anyhow::Result<ChainRuntimeVersion> {
+		Ok(match self.version_mode {
+			RuntimeVersionType::Auto => ChainRuntimeVersion::Auto,
+			RuntimeVersionType::Custom => {
+				let except_spec_version = self.spec_version.ok_or_else(|| {
+					anyhow::Error::msg(format!(
+						"The {}-spec-version is required when choose custom mode",
+						self.prefix
+					))
+				})?;
+				let except_transaction_version = self.transaction_version.ok_or_else(|| {
+					anyhow::Error::msg(format!(
+						"The {}-transaction-version is required when choose custom mode",
+						self.prefix
+					))
+				})?;
+				ChainRuntimeVersion::Custom(except_spec_version, except_transaction_version)
+			},
+			RuntimeVersionType::Bundle => match bundle_runtime_version {
+				Some(runtime_version) => ChainRuntimeVersion::Custom(
+					runtime_version.spec_version,
+					runtime_version.transaction_version,
+				),
+				None => ChainRuntimeVersion::Auto,
+			},
+		})
+	}
+}
+
+/// Create chain-specific set of runtime version parameters.
+#[macro_export]
+macro_rules! declare_chain_runtime_version_params_cli_schema {
+	($chain:ident, $chain_prefix:ident) => {
+		paste::item! {
+			#[doc = $chain " runtime version params."]
+			#[derive(StructOpt, Debug, PartialEq, Eq, Clone, Copy)]
+			pub struct [<$chain RuntimeVersionParams>] {
+				#[doc = "The type of runtime version for chain " $chain]
+				#[structopt(long, default_value = "Bundle")]
+				pub [<$chain_prefix _version_mode>]: RuntimeVersionType,
+				#[doc = "The custom sepc_version for chain " $chain]
+				#[structopt(long)]
+				pub [<$chain_prefix _spec_version>]: Option<u32>,
+				#[doc = "The custom transaction_version for chain " $chain]
+				#[structopt(long)]
+				pub [<$chain_prefix _transaction_version>]: Option<u32>,
+			}
+
+			impl From<[<$chain RuntimeVersionParams>]> for RuntimeVersionParams {
+				fn from(item: [<$chain RuntimeVersionParams>]) -> RuntimeVersionParams {
+					RuntimeVersionParams {
+						prefix: stringify!([<$chain_prefix>]),
+
+						version_mode: item.[<$chain_prefix _version_mode>],
+						spec_version: item.[<$chain_prefix _spec_version>],
+						transaction_version: item.[<$chain_prefix _transaction_version>],
+					}
+				}
+			}
+		}
+	};
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
```

Comment:
I'm still not keen on introducing new structures here. I've looked at all usages and it seems that everywhere (I've been looking at `ConnectionParams` only) we have following code:
```rust
let source_client = ConnectionParams::from(data.source).to_client::<Self::Source>().await?;
```

So it still seems to me that an redundant layer - why not to keep calling `.to_client()` directly on generated structures? :) 
```rust
let source_client = data.source.to_client::<Self::Source>().await?;
```

Don't get me wrong - I'm fine with moving code outside of macro, but this particular thing looks unneeded to me. Like we already had these generated structs before and then we have had `relay_substrate_client::ConnectionParams` with exactly same fields and now we have 3rd version :) Imo it is too much.

The similar situation is with other structs here - e.g. instead of calling `to_keypair` directly on generated struct, we call it like here:
```rust
let target_sign = SigningParams::from(data.target_sign).to_keypair::<Self::Target>()?;
```

TLDR: if there's no need for duplicate structs, I'd avoid adding it. Please don't give up if you have different thoughts on that - I'm ready to change my opinion :)

Reply:
It's true that in these use cases the new structures only add a layer a complexity. Where I did find them more helpful was in the case of the `relay_headers_and_messages`, where they made it easier to work with the CLI supplied params. But also adding some traits there instead of the backing structures might work. I'll give it a try.

Reply:
Ok, if they're really necessary, can we then use `relay_substrate_client::ConnectionParams` instead of this new struct?

Reply:
I removed the backing structures. PTAL on the last commit ! 

Now the problem is that we need a new trait that has to be implemented both by Relay<>Relay bridges and Relay<>Parachains bridges and there is duplication there.

Personally I still prefer the version with the backin structures but both have pros and cons. WDYT ?

Reply:
Yeah - I agree that it looks worse than when structs are used :) Sorry for grumbling about that :) So let's just try to reuse existing structs if possible - `relay_substrate_client::ConnectionParams`, `substrate_relay_helper::TransactionParams`, ...

Reply:
Done. I reverted the chain schema to how it was before and I used backing structures only for the Full2WayBridge . I think it looks better now.